### PR TITLE
NIFI-4583 Restructure nifi-solr-processors

### PR DESCRIPTION
**Pull request by JohannesDaniel**
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by JohannesDaniel** [comment]
Restructured the methods as discussed in JIRA. Additionally, I improved the unit tests for GetSolr

**Comment by JohannesDaniel** [comment]
(tested both processors in local build)

File: nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/src/main/java/org/apache/nifi/processors/solr/SolrUtils.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+public class SolrUtils {
```
Comment (bbende) [inline nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/src/main/java/org/apache/nifi/processors/solr/SolrUtils.java]: Can we move all the static property descriptors to this class, and make SolrUtils consist of only static variables and methods? I think that would fit better with other existing patterns in already in the code-base.

Comment (JohannesDaniel) [reply nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/src/main/java/org/apache/nifi/processors/solr/SolrUtils.java]: Kojis idea was to define createSolrClient as non-static to facilitate unit-testing. I now have defined everything as static.

Comment (JohannesDaniel) [reply nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/src/main/java/org/apache/nifi/processors/solr/SolrUtils.java]: @bbende any other comments? As soon as this pull request is merged, I can create a pull request for NIFI-4516

File: nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/src/main/java/org/apache/nifi/processors/solr/SolrProcessor.java [hunk 2]
```diff
@@ -34,39 +31,33 @@
...
+    public static PropertyDescriptor SOLR_TYPE = SolrUtils.SOLR_TYPE;
+    public static PropertyDescriptor COLLECTION = SolrUtils.COLLECTION;
+    public static PropertyDescriptor JAAS_CLIENT_APP_NAME = SolrUtils.JAAS_CLIENT_APP_NAME;
+    public static PropertyDescriptor SSL_CONTEXT_SERVICE = SolrUtils.SSL_CONTEXT_SERVICE;
+    public static PropertyDescriptor SOLR_SOCKET_TIMEOUT = SolrUtils.SOLR_SOCKET_TIMEOUT;
+    public static PropertyDescriptor SOLR_CONNECTION_TIMEOUT = SolrUtils.SOLR_CONNECTION_TIMEOUT;
+    public static PropertyDescriptor SOLR_MAX_CONNECTIONS = SolrUtils.SOLR_MAX_CONNECTIONS;
+    public static PropertyDescriptor SOLR_MAX_CONNECTIONS_PER_HOST = SolrUtils.SOLR_MAX_CONNECTIONS_PER_HOST;
+    public static PropertyDescriptor ZK_CLIENT_TIMEOUT = SolrUtils.ZK_CLIENT_TIMEOUT;
+    public static PropertyDescriptor ZK_CONNECTION_TIMEOUT = SolrUtils.ZK_CONNECTION_TIMEOUT;
```
Comment (ijokarumawak) [inline nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/src/main/java/org/apache/nifi/processors/solr/SolrProcessor.java]: Do we still need these values defined here? Static import would be enough.

File: nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/src/main/java/org/apache/nifi/processors/solr/SolrProcessor.java [hunk 3]
```diff
@@ -205,47 +125,7 @@ public final void closeClient() {
...
-            final Integer zkClientTimeout = context.getProperty(ZK_CLIENT_TIMEOUT).asTimePeriod(TimeUnit.MILLISECONDS).intValue();
-            final Integer zkConnectionTimeout = context.getProperty(ZK_CONNECTION_TIMEOUT).asTimePeriod(TimeUnit.MILLISECONDS).intValue();
-
-            CloudSolrClient cloudSolrClient = new CloudSolrClient(solrLocation, httpClient);
-            cloudSolrClient.setDefaultCollection(collection);
-            cloudSolrClient.setZkClientTimeout(zkClientTimeout);
-            cloudSolrClient.setZkConnectTimeout(zkConnectionTimeout);
-            return cloudSolrClient;
-        }
+        return SolrUtils.createSolrClient(context, solrLocation);
```
Comment (ijokarumawak) [inline nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/src/main/java/org/apache/nifi/processors/solr/SolrProcessor.java]: About the static or non-static method for unit testing, this createSolrClient method at SolrProcessor can be overridden at test case, so SolrUtils can have everything as static. It looks good, sorry for the confusion.

File: nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/src/main/java/org/apache/nifi/processors/solr/SolrProcessor.java (hunk 2 above)
Comment (JohannesDaniel) [reply nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/src/main/java/org/apache/nifi/processors/solr/SolrProcessor.java]: Hi, thx for response. Changing this to static imports means that several PropertyDescriptors have to be imported in classes extending SolrProcessor as well. My idea was to add the PropertyDescriptors as values that they can be used in the same way they were used before. However, I can change this.

Comment (JohannesDaniel) [reply nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/src/main/java/org/apache/nifi/processors/solr/SolrProcessor.java]: This also affects all JUnit tests for GetSolr and PutSolrContentStream. E. g. all lines like
runner.setProperty(GetSolr.SOLR_TYPE, GetSolr.SOLR_TYPE_CLOUD.getValue());
have to be changed...

Comment (JohannesDaniel) [reply nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/src/main/java/org/apache/nifi/processors/solr/SolrProcessor.java]: I suggest to leave it as it is...

Comment (ijokarumawak) [reply nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/src/main/java/org/apache/nifi/processors/solr/SolrProcessor.java]: I think this PR's goal is restructuring existing code by centralizing common implementations into SolrUtils. Nothing wrong to change existing code. Do you envision when you add new PropertyDescriptor, you prefer to add it in SolrUtils and also in SolrProcessor? If you still prefer leave it as it is, please let me know. I just wanted to confirm before I merge this. Thanks!

Comment (JohannesDaniel) [reply nifi-nar-bundles/nifi-solr-bundle/nifi-solr-processors/src/main/java/org/apache/nifi/processors/solr/SolrProcessor.java]: You are right! I would add new PropertyDescriptors  only to SolrProcessor if the are only supposed to be used for processors.

**Comment by JohannesDaniel** [comment]
I made TestGetSolr a bit more simpler by adding a method for the generic creation of TestRunners. I moved all remaining PropertyDescriptors from SolrProcessor to SolrUtils as they will be needed also for e. g. ControllerServices.

**Comment by JohannesDaniel** [comment]
I replaced the values by static imports

**Comment by ijokarumawak** [comment]
@JohannesDaniel Thanks for the updates. Now it LGTM, +1. Merging in. Thank you for your contribution!

**Comment by JohannesDaniel** [comment]
@ijokarumawak not too important, but do you know why this commit is not linked to my account? it is linked to "U-WOODMARK\johannes.peter"

**Comment by ijokarumawak** [comment]
@JohannesDaniel Would you check your working environment setting, especially email address? I just squashed commits into one and pushed into master branch, didn't modify author info.
https://help.github.com/articles/setting-your-commit-email-address-in-git/

**Comment by JohannesDaniel** [comment]
Ah I see. Sorry for this. Will we leave it as it (for this commit) is or can we fix this?

**Comment by ijokarumawak** [comment]
@JohannesDaniel Technically, we can [amend old commit info](https://stackoverflow.com/questions/3042437/change-commit-author-at-one-specific-commit). However, I think it requires a force push, which may lose commits if coming at the same time, and for every local branch people cloned to be updated. So if we decide to do that operation, we need to announce that to NiFi dev ML at least. Please initiate such conversation at NiFi dev ML if needed. Thank you.

**Comment by ijokarumawak** [comment]
So, I'd prefer leaving as it is if it's not too critical.

**Pull request closed without merging**